sha256-avx2/utilsx8.c

```diff
@@ -5,121 +5,142 @@
-                               tree_idx + (idx_offset[j] >> (heights[offset-1] + 1)));
+            /*
+             * Check if we're at a left child; if so, stop going up the stack
+             * Exception: if we've reached the end of the tree, keep on going
+             * (so we combine the last 8 nodes into the one root node in three
+             * more iterations)
+             */
+            if ((internal_idx & 1) == 0 && idx < max_idx) {
+                break;
             }
-            /* Hash the top-most nodes from the stack together. */
-            thashx8(stackx8 + 0*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 1*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 2*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 3*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 4*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 5*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 6*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 7*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 0*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 1*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 2*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 3*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 4*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 5*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 6*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 7*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N, 2, pub_seed, tree_addrx8);
-            offset--;
-            /* Note that the top-most node is now one layer higher. */
-            heights[offset - 1]++;
 
-            /* If this is a node we need for the auth path.. */
+            /* Ok, we're at a right node (or doing the top 3 levels) */
+            /* Now combine the left and right logical nodes together */
+
+            /* Set the address of the node we're creating. */
+            int j;
+            internal_idx_offset >>= 1;
             for (j = 0; j < 8; j++) {
-                if (((leaf_idx[j] >> heights[offset - 1]) ^ 0x1) == tree_idx) {
-                    memcpy(auth_pathx8 + j*tree_height*SPX_N + heights[offset - 1]*SPX_N,
-                           stackx8 + j*(tree_height + 1)*SPX_N + (offset - 1)*SPX_N, SPX_N);
-                }
+                set_tree_height(tree_addrx8 + j*8, h + 1);
+                set_tree_index(tree_addrx8 + j*8,
+                     (8/2) * (internal_idx&~1) + j - left_adj + internal_idx_offset );
             }
+            unsigned char *left = &stackx8[h * 8 * SPX_N];
+            thashx8( &current[0 * SPX_N],
```

Comment:
Perhaps recall that the logical node we're computing now is the combination of the old current logical node (which on all but the last three levels is on the right) and the previously saved logical node on the left.  The first four nodes of our new logical node are pairwise hashes within the left logical node and the last four are pairwise hashes within the old current.

On the second-to-last-level we only need to compute 4 nodes and our 8 input nodes are all in `current`.  Out of convenience we do the same computation, but that shifts the output nodes 4 nodes to the right (and so we need `left_adj=4`.)